traitlets/traitlets.py

```diff
@@ -1412,6 +1412,34 @@ def has_trait(self, name):
         """Returns True if the object has a trait with the specified name."""
         return isinstance(getattr(self.__class__, name, None), TraitType)
 
+    def trait_values(self, **metadata):
+        """A ``dict`` of trait names and their values.
+
+        The metadata kwargs allow functions to be passed in which
+        filter traits based on metadata values.  The functions should
+        take a single value as an argument and return a boolean.  If
+        any function returns False, then the trait is not included in
+        the output.  If a metadata key doesn't exist, None will be passed
+        to the function.
+
+        Returns
+        -------
+        A ``dict`` of trait names and their values.
+
+        Notes
+        -----
+        Trait values are retrieved via ``getattr``, any exceptions raised
+        by traits or the operations they may trigger will result in the
+        absence of a trait value in the result ``dict``.
+        """
+        values = {}
+        for name in self.trait_names(**metadata):
+            try:
+                values[name] = getattr(self, name)
+            except:
+                pass
```

Comment:
I'm wasn't entirely sure of how to handle this, because traits which have no assigned value and which have no default value, will raise when requested via `getattr`. However, the intuitive expectation for such traits is that they simply be excluded from the returned `dict`. I would have just returned a copy of `HasTraits._trait_values` however, this would mean that traits which have valid default values would not be represented.

Perhaps `values[name] = error` if an exception is encountered?